arch: X86: update x86 arch porting with new acpica interface

najumon1980 opened the pull request:
update x86 arch porting with new acpica interface. As part of this the old x86 acpi implementation has been removed and updated test apps and other board configurations with new  acpica interface support.

najumon1980 committed:
lib: acpi: add support for MAD table and DMAR table

add support for retrieve MAD and DMAR table information. Provided
two new interface namely acpi_dmar_entry_get() and acpi_drhd_get()
for retrieve DMA Remapping Reporting and DMA-remapping hardware
unit definition (DRDH).

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
arch: x86: update with new acpica lib interface

The old acpi implimentation is replaced with acpica interface
and updated x86 arch porting with the new interface.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
test: arch: x86: update with new acpica interface

modified existing x86 tests app with the new acpica interface

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
board: qemu_x86: update with new acpica interface

update qemu, up_squared, acrn and ehl platform board config with
new acpica interface

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 commented:
Will add board/soc update for other x86 platforms once tested

najumon1980 commented:
Few CI job are failed and this will resolve once acpica PR get merged.

najumon1980 commented:
@andyross could you please have a look into the changes ?

tbursztyka commented on include/zephyr/acpi/acpi.h:
```diff
@@ -117,9 +126,30 @@ int acpi_device_type_get(ACPI_RESOURCE *res);
  * @return return 0 on success or error code
  */
-int acpi_table_get(char *signature, int inst, void **acpi_table);
+int acpi_get_madt(int type, struct acpi_subtable_header **tables, int *num_inst, int max_inst);
```
acpi sub-tables are all contiguous and persistent. Why one would need to allocate another place-holder, with the risk of over-allocating (memory loss) and under-allocating (information loss).

Just follow what has been done so far.

tbursztyka commented on lib/acpi/Kconfig:
```diff
@@ -52,3 +52,9 @@ config ACPI_HID_LEN_MAX
 	help
 	  Size of HID string.
+
+config ACPI_MADT_LAPIC_MAX
```
So this type of options would be totally useless.

tbursztyka commented on lib/acpi/acpi.c:
```diff
@@ -224,8 +223,8 @@ static ACPI_STATUS acpi_enable_pic_mode(void)
 
-static int acpi_get_irq_table(struct acpi *bus, char *bus_name,
-				ACPI_PCI_ROUTING_TABLE *rt_table, uint32_t rt_size)
+static int acpi_get_irq_table(struct acpi *bus, char *bus_name, ACPI_PCI_ROUTING_TABLE *rt_table,
```
why changing this?

tbursztyka commented on lib/acpi/acpi.c:
```diff
@@ -491,8 +492,7 @@ int acpi_current_resource_free(ACPI_RESOURCE *res)
 
-int acpi_get_irq_routing_table(char *bus_name,
-			       ACPI_PCI_ROUTING_TABLE *rt_table, size_t rt_size)
+int acpi_get_irq_routing_table(char *bus_name, ACPI_PCI_ROUTING_TABLE *rt_table, size_t rt_size)
```
same here, why?

tbursztyka commented on lib/acpi/acpi.c:
```diff
@@ -633,17 +633,137 @@ int acpi_table_get(char *signature, int inst, void **acpi_table)
 		status = acpi_early_init();
 		if (status) {
 			LOG_ERR("ACPI early int failed");
```
s/int/init

tbursztyka commented on lib/acpi/acpi.c:
```diff
@@ -633,17 +633,137 @@ int acpi_table_get(char *signature, int inst, void **acpi_table)
+	return (void *)table;
+}
+
+int acpi_get_madt(int type, struct acpi_subtable_header **tables, int *num_inst, int max_inst)
```
actually the function is not even doing what its name means: it is not returning the madt table, but fills in an array of pointer of all (as long as allocation fits, see previous comment), sub-table pointer. 

"acpi_get_madt_entry" would somehow be more semantically right.

tbursztyka commented on lib/acpi/acpi.c:
```diff
@@ -633,17 +633,137 @@ int acpi_table_get(char *signature, int inst, void **acpi_table)
+	return 0;
+}
+
+static int acpi_dmar_get(enum AcpiDmarType type, struct acpi_subtable_header **tables,
```
same here, this does not return dmar table (and has the same allocation issue).

tbursztyka commented on modules/acpica/Kconfig:
```diff
@@ -6,4 +6,10 @@ config ACPI
 
 menu "ACPI driver support"
 
+config ACPI_FULL_SUPPORT
```
nah, either we support acpi or we don't. Let's not add intermediate support (which is cryptic as well: what means CONFIG_ACPI alone vs CONFIG_ACPI_FULL_SUPPORT then?)

tbursztyka replied:
So basically, it's the difference between getting acpica that enables only basic ACPI tables and acpica that enables the same plus DSDT support? 

In that case, name this option ACPI_DSDT_SUPPORT, it's clearer imo.

tbursztyka reviewed (changes requested):
Instead of porting, you change the logic. Which is premature as well: for instance, we are not going to support more than one ioapic anytime soon (if ever, actually).

plus various unnecessary changes etc..

najumon1980 commented:
@tbursztyka update as per your review comment and also responded to your few comments

jhedberg commented:
@najumon1980 looks like there's a bunch of build failures:
```
/__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c: In function 'AcpiOsReadMemory':
/__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c:390:31: error: implicit declaration of function 'sys_read64'; did you mean 'sys_read32'? [-Werror=implicit-function-declaration]
  390 |         *((UINT64 *) Value) = sys_read64 (Address);
      |                               ^~~~~~~~~~
      |                               sys_read32
/__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c: In function 'AcpiOsWriteMemory':
/__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c:441:9: error: implicit declaration of function 'sys_write64'; did you mean 'sys_write32'? [-Werror=implicit-function-declaration]
  441 |         sys_write64 ((UINT64) Value, Address);
      |         ^~~~~~~~~~~
      |         sys_write32
/__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c: In function 'AcpiOsGetRootPointer':
/__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c:778:13: error: cast from pointer to integer of different size [-Werror=pointer-to-int-cast]
  778 |     return ((ACPI_PHYSICAL_ADDRESS) efi_get_acpi_rsdp ());
      |             ^
```

najumon1980 commented:
> @najumon1980 looks like there's a bunch of build failures:
> 
> ```
> /__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c: In function 'AcpiOsReadMemory':
> /__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c:390:31: error: implicit declaration of function 'sys_read64'; did you mean 'sys_read32'? [-Werror=implicit-function-declaration]
>   390 |         *((UINT64 *) Value) = sys_read64 (Address);
>       |                               ^~~~~~~~~~
>       |                               sys_read32
> /__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c: In function 'AcpiOsWriteMemory':
> /__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c:441:9: error: implicit declaration of function 'sys_write64'; did you mean 'sys_write32'? [-Werror=implicit-function-declaration]
>   441 |         sys_write64 ((UINT64) Value, Address);
>       |         ^~~~~~~~~~~
>       |         sys_write32
> /__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c: In function 'AcpiOsGetRootPointer':
> /__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c:778:13: error: cast from pointer to integer of different size [-Werror=pointer-to-int-cast]
>   778 |     return ((ACPI_PHYSICAL_ADDRESS) efi_get_acpi_rsdp ());
>       |             ^
> ```

This need following acpica PR need to merge and update the west.yml accordingly. 
https://github.com/acpica/acpica/pull/882

jhedberg commented:
> This need following acpica PR need to merge and update the west.yml accordingly. [acpica/acpica#882](https://github.com/acpica/acpica/pull/882)

Please submit a PR to our ACPICA mirror (which will not be merged - we’ll sync the tree when the ACPICA upstream PR is merged), and follow up these instructions: https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-changes-to-existing-modules

That way we can verify that CI passes with the combination of the ACPICA PR and this PR.

tbursztyka reviewed (dismissed):
Fix the Kconfig option as suggested. 

Rest lgtm

tbursztyka commented on include/zephyr/arch/x86/x86_acpi_osal.h:
```diff
@@ -0,0 +1,25 @@
+
+uint32_t bios_acpi_rsdp_get(void);
+
+#if defined(CONFIG_X86_EFI)
```
No you can't assume that once EFI is enabled, the system will always get it successfully from EFI.

Look at the original code you removed in arch/x86/core/acpic.c: is efi fails to get rsdp, it falls back to legacy lookup.

tbursztyka commented on include/zephyr/acpi/acpi.h:
```diff
@@ -19,6 +19,22 @@ struct acpi_dev {
+};
+
+struct acpi_mcfg {
+	struct acpi_table_header Header;
```
why the capital letter ? 

Unless we use an externally declared structures/variable/etc... (so from hal, lib etc...), we do not use came case in Zephyr. So here, just use small letter.

tbursztyka commented on tests/kernel/smp/src/main.c:
```diff
@@ -372,7 +372,7 @@ ZTEST(smp, test_coop_resched_threads)
 	 * will not get scheduled
 	 */
-	spawn_threads(K_PRIO_COOP(10), num_threads, !EQUAL_PRIORITY,
+	spawn_threads(K_PRIO_COOP(MAX_NUM_THREADS), num_threads, !EQUAL_PRIORITY,
```
how does this relate to ACPICA? (as above in the smp's prj.conf)

Make it its own commit, and have a proper commit message explaining the change.

tbursztyka commented on tests/arch/x86/info/src/acpi.c:
```diff
@@ -4,145 +4,47 @@
-			struct acpi_cpu *cpu = z_acpi_get_cpu(i);
-			printk("\tCPU #%d: APIC ID 0x%02x\n", i, cpu->apic_id);
+		for (i = 0; i < CONFIG_MP_MAX_NUM_CPUS; ++i) {
+			printk("\tCPU #%d: APIC ID 0x%02x\n", i, lapic[i]->Id);
```
No this is bogus: it can happen that a defined CPU in madt might not be enabled at all,  so we don't want to display those. Thus the logic above in acpi() that first counts the valid CPUs. Don't change that.

tbursztyka replied:
Nevermind this comment, I forgot that z_acpi_get_cpu() does the check for disablement. 

and yeah, the above logic is used only to verify that we got rsdt/madt.

tbursztyka commented on tests/arch/x86/info/src/acpi.c:
```diff
@@ -4,145 +4,47 @@
+	       dmar_id->bits.function);
 }
 
-static void vtd_drhd_info(struct acpi_drhd *drhd)
```
why removing this this? Keep it.

tbursztyka commented on tests/arch/x86/info/src/acpi.c:
```diff
@@ -4,145 +4,47 @@
+	union acpi_dmar_id *dmar_id;
+	int num_inst;
 
-	switch (dev_scope->type) {
```
why removing this? Keep it.

tbursztyka reviewed (changes requested):
Still quite some issues.

First: have proper commit message. Copy/pasting the commit title in the message is not a proper message. Unless the change is supper trivial, you need to give a little explanation on what the commit is doing.

Second: when you port code to a new interface, you'll need to port existing logic and **not** removing it. See my comments in tests/arch  etc...

Also, let's wait now that ACPICA gets the PR 882 applied, let's update the west manifest and then let's see if CI in this present PR passes.

najumon1980 committed:
modules: lib: acpica: update west to fetch acpica PR 882

update west to fetch acpica PR 882 for running CI job. This is a temp
commit to verify CI job and will update with latest commit once acpica
PR get merged into mainline.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>
